### ISSUE TITLE
network config: static route and additional / second public ip

### ISSUE THREAD

**Issue by pwFoo** [rancher/os#400]
To use RancherOS in our special network setup I have to
1. add a second ip address to the first nic (two public ip addresses, not tested with rancheros yet...)
2. set a static route during boot to the default gw

Without the static route RancherOS won't be able to access the internet.

RancherOS network config:
ip address: 
<IP-1>/32
<IP-2>/32

routes
<GW-IP> dev eth0  scope link                                                                                                                                                         
default via <GW-IP> dev eth0 

Because of the /32 ip address we need a route to the default gw to reach it.

**Comment by imikushin** [comment]
Currently (v0.3.1) RancherOS doesn't have routing configuration in cloud-config. What you can do though is start a privileged container with --net=host and perform all the necessary network fine tuning in it, with `ip` commands. You can also put that container into your cloud-config and thus have it run on boot. E.g. like this:
```yml
#cloud-config
rancher:
  system_containers:
    my-routing-config:
      image: my-routing-config/image:foo
      privileged: true
      net: host
```

**Comment by pwFoo** [comment]
Thanks, I'll try this. 

Will the container executed early enough to setup the route to the default gateway? Could be a problem if the internet connection will be up to late (ntpd or docker related tasks)?

With another machine I experienced some problems with delayed ssh access (with kvm vm it's fast, with a mini server it takes 3+ minutes). And sometimes it seems the server gets no dhcp ip address (static lease), because I can't ping / ssh for minutes. After a reboot dhcp is working again and after 3+ minutes I can login via ssh...

**Comment by imikushin** [comment]
Your container image can be [pre-packed](http://os.docs.rancher.com/docs/configuration/prepacking-docker-images/) so system-docker doesn't try to download it before running it. As you're trying to tweak network settings, you should definitely use system-docker, because you can run a system service before network becomes available to (user) docker daemon. 

That's why I see I've made a mistake in my previous message: you need to use `services` instead of `system_containers` in your cloud-config, because system containers are loaded by system-docker on boot from initrd, and service images can be preloaded at a later stage. You'll also need to label your container as system and non-detach (batch job on system start). So, the code sample should look like this:
```yml
#cloud-config
rancher:
  services:
    my-routing-config:
      image: my-routing-config/image:foo
      labels:
        io.rancher.os.detach: false
        io.rancher.os.scope: system
      privileged: true
      net: host
      links:
      - network
```
Link to network service if you need basic network configuration from cloud-config to be applied before running your container.

**Comment by pwFoo** [comment]
Thank's I'll test it soon. 
So ip / route configuration will be added to rancheros cloud config in the future (beta / stable release)?

**Comment by imikushin** [comment]
That is quite possible.

**Comment by pwFoo** [comment]
Could I set routes that way?
```
 write_files:
 - path: /opt/rancher/bin/start.sh
    permissions: 0755
    owner: root
    content: |
      #!/bin/bash
      echo "I'm doing things on start"
```
Or will network commands like ip route not available during boot?

**Comment by imikushin** [comment]
This is quite possible. Keep in mind that `/opt/rancher/bin/start.sh` is run by the console service on start. Console starts after basic network config (from cloud-init) has been applied. So in this case you might want to link your services to console. 

I'd choose a dedicated container (as shown above) because of precise dependency control: you can label your custom-routing-config as detach=false and this makes sure it has completed before starting dependent services.

**Comment by TotalLag** [comment]
Hello, I am also in the same situation and need to pre-configure network routes before it can be online.  I only have access to a KVM, so if I install via ISO, there's no way for me to login since it is looking for a SSH key.

I have tried both write_files and a privileged container without success (probably not setting something up right for the container) but this could be easily solved if it were possible to pass a users parameter in the cloud-config.yml like in https://coreos.com/os/docs/latest/cloud-config.html#users to manually create a user so we can login to configure networking.

**Comment by pwFoo** [comment]
Login with password would be great to administrate RancherOS VMs :+1: 
Should be possible and documented how to enable login with password.

**Comment by deniseschannon** [comment]
@pwFoo @TotalLag Currently, there is no ability to add users to RancherOS. There is a separate issue to track that feature. https://github.com/rancher/os/issues/263

**Comment by pwFoo** [comment]
Back to othe network configs by a service container.

Wouldn't it a problem to manage network settings from a service container because RancherOS could be unreachable if something goes wrong...
For example eth0 is configured by RancherOS and the service container is used to configure eth1 and eth2. During tests after network restart eth0 was down / without an ip address.

I don't like two different points (RancherOS and an additional service container) to manage interfaces. Maybe it could be fine if only the additional (physical) interface is mapped to the container instead all interfaces?

**Comment by pwFoo** [comment]
Hello @imikushin 

as you posted at [issuecomment-115158109](https://github.com/rancher/os/issues/400#issuecomment-115158109) a config container should be the best way.
Should it based on a RancherOS / os-service image? Like console? WOuld be great to understand how to hook into the RancherOS / os-services.


Regards

**Comment by pwFoo** [comment]
Closed, because related to #660.
